documentation/jetty-documentation/src/main/asciidoc/programming-guide/arch-threads.adoc

```diff
@@ -0,0 +1,200 @@
+Deciding between these two modes depends on whether there is a free thread immediately available to take over production, and this is captured by the `org.eclipse.jetty.util.thread.TryExecutor` interface.
+
+An implementation of `TryExecutor` could be asked whether it can immediately allocate a free thread to run a task, as opposed to a normal `Executor`, that can only queue the task in the hope that there will be a thread in the near future that could run the task.
+
+The concept of task consumption modes, coupled with `Invocable` tasks that expose their behavior, coupled with a `TryExecutor` that guarantees whether production can be immediately taken over are captured by the default Jetty execution strategy, named `org.eclipse.jetty.util.thread.AdaptiveExecutionStrategy`.
+
+[NOTE]
+====
+`AdaptiveExecutionStrategy` was previously named `EatWhatYouKill`, named after a hunting proverb in the sense that one should produce (kill) only what it consumes (eats).
+====
+
+[[pg-arch-threads-thread-pool]]
+==== Thread Pool
+Jetty's xref:pg-arch-threads[threading architecture] requires a more sophisticated thread pool than what offered by Java's `java.util.concurrent.ExecutorService`.
+
+Jetty's default thread pool implementation is link:{javadoc-url}/org/eclipse/jetty/util/thread/QueuedThreadPool.html[`QueuedThreadPool`]
+`QueuedThreadPool` integrates with the xref:pg-arch-bean[Jetty component model], implements `Executor`, provides a `TryExecutor` implementation (discussed in the xref:pg-arch-threads-execution-strategy-adaptive[adaptive execution strategy section]), and supports xref:pg-arch-threads-thread-pool-virtual-threads[virtual threads] (introduced as a preview feature in Java 19).
+
+`QueuedThreadPool` can be configured with a `maxThreads` value.
+
+However, some of the Jetty components (such as the xref:pg-arch-io-selector-manager[selectors]) permanently steal threads for their internal use, or rather `QueuedThreadPool` leases some threads to these components.
+These threads are reported by `QueuedThreadPool.leasedThreads` and are not available to run application code.
+
+`QueuedThreadPool` can be configured with a `reservedThreads` value.
+This value represents the maximum number of threads that can be reserved and used by the `TryExecutor` implementation.
+
+
+[[pg-arch-threads-thread-pool-virtual-threads]]
+===== Virtual Threads
+Virtual threads have been introduced in Java 19 as a preview feature.
+
+NOTE: In Java versions where virtual threads are a preview feature, remember to add `+--enable-preview+` to the command line options to use virtual threads.
+
+`QueuedThreadPool` can be configured to use virtual threads via:
+
+[source,java,indent=0]
+----
+include::{doc_code}/org/eclipse/jetty/docs/programming/QueuedThreadPoolDocs.java[tags=virtual-threads]
+----
```

Comment:
```suggestion
----
[NOTE]
====
If virtual thread are enabled in `QueuedThreadPool`, then the heuristic for determining the number of reserved threads will return 0, so that a virtual thread will always be used for blocking tasks.   If a deployment wishes to use enable the use of Execute-Produce-Consume strategy for blocking tasks, then a specific number of reserved threads must be configured. 
====
```

Reply:
Modified and expanded, but done.